src/vm/arm/asmhelpers.S

```diff
@@ -1450,3 +1450,42 @@ DelayLoad_Helper\suffix:
         NESTED_END OnHijackTripThread, _TEXT
 #endif
 
+// ------------------------------------------------------------------
+// The following helper will access ("probe") a word on each page of the stack
+// starting with the page right beneath sp down to the one pointed to by r4.
+// The procedure is needed to make sure that the "guard" page is pushed down below the allocated stack frame.
+// The call to the helper will be emitted by JIT in the function/funclet prolog when large (larger than 0x3000 bytes) stack frame is required.
+// On entry:
+//   r4 - points to the lowest address on the stack frame being allocated (i.e. [InitialSp - FrameSize])
+//   sp - points to some byte on the last probed page
+// On exit:
+//   r4 - is preserved
+//   r5 - is not preserved
+//
+// NOTE: this helper will probe at least one page below the one pointed to by sp.
+#define PAGE_SIZE 0x1000
+
+        NESTED_ENTRY JIT_StackProbe, _TEXT, NoHandler
+        PROLOG_PUSH "{fp,lr}"
+        PROLOG_STACK_SAVE fp
+
+        // NOTE: we can not use sp for computations below due to:
+        //   1. 'SUB sp' immediate must be in the range -508 to +508;
+        //   2. 'ORR' operands registers must be in the range r0-r7.
+        sub r5, sp, PAGE_SIZE            // r5 points to some byte on the first unprobed page
+
+        orr r5, (PAGE_SIZE - 1) & 0xFF
```

Comment:
I do not think it is a good idea to ever have misaligned SP. It has a high chance of being misidentified as corrupted process state. Also, the probing loads below will be misaligned - not the best thing to do for performance.

Reply:
Please fix that for x86/x64 as well. I do know that it is a problem on x86/x64 - we used a similar misaligned esp trick in the runtime a long time, but we have to get rid of it because of it was causing all sorts of problems.

Reply:
@jkotas Thanks for bringing this up! I didn't know that it was causing problems in the runtime before. Will fix.

I know that on Arm64 the spec requires for SP always be 16-byte aligned - for x86/x64/arm I could not find such requirement - the alignment requirement there only exists for call sites.

Reply:
@echesakov have you looked at what native compilers do for probing on arm32/army64? They should have to solve a similar problem...

Reply:
@AndyAyersMS Yes, I have.

**VC++ (win-arm)** produces
```
00ca1ce0 e92d4830 push        {r4,r5,r11,lr}
00ca1ce4 f10d0b08 add         r11,sp,#8
00ca1ce8 f7ffff52 bl          _chkstk!__security_push_cookie (00ca1b90)
00ca1cec f2440401 mov         r4,#0x4001
00ca1cf0 f7ffff74 bl          _chkstk!__chkstk (00ca1bdc)
00ca1cf4 ebad0d04 sub         sp,sp,r4
```

**VC++ (win-arm64)** produces
```
00007ff7`f6c91b90 a9bf7bfd stp         fp,lr,[sp,#-0x10]!
00007ff7`f6c91b94 910003fd mov         fp,sp
00007ff7`f6c91b98 97fffe6a bl          _chkstk!__security_push_cookie (00007ff7`f6c91540)
00007ff7`f6c91b9c d282000f mov         x15,#0x1000
00007ff7`f6c91ba0 97fffe88 bl          _chkstk!__chkstk (00007ff7`f6c915c0)
00007ff7`f6c91ba4 cb2f73ff sub         sp,sp,x15,lsl #4
```

I try to do the same using `JIT_StackProbe` helper. 

The issue I was talking about is that `CodeGen::genAllocLclFrame` on arm64 only does stack probing. So we would insert instruction that saves `lr` on stack after the call to the helper - this is a main difference between us and VC++.

Reply:
We could create yet another stack frame type for arm64 where, if a call to chkstk is required, save fp/lr first, then call chksk, then do the rest of the frame setup. We have a frame type a little like this for GS cookie + localloc case; perhaps those could be re-used, if necessary.

Reply:
If that can be done, it sounds like the best solution to me.

Reply:
@BruceForstall I will take a look at this option. I am afraid that the consistency of using stack probing helpers on all platforms might not be worth complexity added to the JIT by introducing one more stack frame type. We'll see.

Reply:
btw, while I stated we _could_ do this, it's not clear that it's worth the effort or complexity, so I wouldn't hold up this change for that investigation or work, and wouldn't necessarily want to spent much time on it, either.